src/core/desktopentrymonitor.cpp

```diff
@@ -0,0 +1,188 @@
+
+namespace {
+QS_LOGGING_CATEGORY(logDesktopMonitor, "quickshell.desktopentrymonitor", QtWarningMsg);
+}
+
+DesktopEntryMonitor::DesktopEntryMonitor(QObject* parent): QObject(parent) {
+	this->watcher = new QFileSystemWatcher(this);
+	this->debounceTimer = new QTimer(this);
+	this->debounceTimer->setSingleShot(true);
+	this->debounceTimer->setInterval(500); // 500ms debounce
+
+	// Initialize XDG desktop paths
+	this->initializeDesktopPaths();
+
+	// Setup connections
+	connect(
+	    this->watcher,
+	    &QFileSystemWatcher::directoryChanged,
+	    this,
+	    &DesktopEntryMonitor::onDirectoryChanged
+	);
+	connect(
+	    this->watcher,
+	    &QFileSystemWatcher::fileChanged,
+	    this,
+	    &DesktopEntryMonitor::onFileChanged
+	);
+	connect(this->debounceTimer, &QTimer::timeout, this, &DesktopEntryMonitor::processChanges);
+
+	// Start monitoring
+	this->startMonitoring();
+}
+
+void DesktopEntryMonitor::initializeDesktopPaths() {
+	this->desktopPaths = DesktopUtils::getDesktopDirectories();
+}
+
+void DesktopEntryMonitor::startMonitoring() {
+	for (const QString& path: this->desktopPaths) {
+		if (QDir(path).exists()) {
+			qCDebug(logDesktopMonitor) << "Monitoring desktop entry path:" << path;
+			this->addDirectoryHierarchy(path);
+		}
+	}
+}
+
+void DesktopEntryMonitor::addDirectoryHierarchy(const QString& dirPath) {
+	this->scanAndWatch(dirPath);
+}
+
+void DesktopEntryMonitor::scanAndWatch(const QString& dirPath) {
+	QDir dir(dirPath);
+	if (!dir.exists()) return;
+
+	// Add directory to watcher
+	if (this->watcher->addPath(dirPath)) {
+		qCDebug(logDesktopMonitor) << "Added directory to watcher:" << dirPath;
+	}
+
+	// Add .desktop files
+	for (const auto& entry: dir.entryInfoList({"*.desktop"}, QDir::Files)) {
+		auto path = entry.absoluteFilePath();
+		if (!this->watchedFiles.contains(path)) {
+			if (this->watcher->addPath(path)) {
+				this->fileTimestamps[path] = entry.lastModified();
+				this->watchedFiles.insert(path);
+				qCDebug(logDesktopMonitor) << "Monitoring file:" << path;
+			}
+		}
+	}
+
+	// Recurse into subdirs
+	for (const auto& sub: dir.entryList(QDir::Dirs | QDir::NoDotAndDotDot)) {
+		this->scanAndWatch(dir.absoluteFilePath(sub));
+	}
+}
+
+void DesktopEntryMonitor::onDirectoryChanged(const QString& path) {
+	qCDebug(logDesktopMonitor) << "Directory changed:" << path;
+	QDir dir(path);
+
+	// Check if directory still exists - handle removal/unmounting
+	if (!dir.exists()) {
+		qCDebug(logDesktopMonitor) << "Directory no longer exists, cleaning up:" << path;
+		this->watcher->removePath(path);
+
+		// Remove all watched files from this directory and its subdirectories
+		for (auto it = this->watchedFiles.begin(); it != this->watchedFiles.end();) {
+			const QString& watchedFile = *it;
+			if (watchedFile.startsWith(path + "/") || watchedFile == path) {
+				this->watcher->removePath(watchedFile);
+				this->fileTimestamps.remove(watchedFile);
+				this->queueChange(ChangeEvent::Removed, watchedFile);
```

Comment:
The debounce is not used in a logical way. If you are going to debounce, it should guard re-scanning directories, not model updates.